addons/mail/models/res_users.py

```diff
@@ -35,12 +35,30 @@ class Users(models.Model):
         "Only internal user can receive notifications in Odoo",
     )]
 
-    @api.depends('share')
+    @api.depends('share', 'groups_id')
     def _compute_notification_type(self):
+        inbox_group_id = self.env['ir.model.data']._xmlid_to_res_id('mail.group_mail_notification_type_inbox')
+
+        self.filtered_domain([
+            ('groups_id', 'in', inbox_group_id), ('notification_type', '!=', 'inbox')
+        ]).notification_type = 'inbox'
+        self.filtered_domain([
+            ('groups_id', 'not in', inbox_group_id), ('notification_type', '=', 'inbox')
+        ]).notification_type = 'email'
+
+        # Special case: internal users with inbox notifications converted to portal must be converted to email users
+        self.filtered_domain([('share', '=', True), ('notification_type', '=', 'inbox')]).notification_type = 'email'
```

Comment:
Cannot we just do that?
```suggestion
        users_with_notification_inbox = self.filtered_domain([
            ('groups_id', 'in', inbox_group_id), ('share', '=', False)
        ])
        users_with_notification_inbox.notification_type = 'inbox'
        (self - users_with_notification_inbox).notification_type = 'email'
```